core/src/main/java/org/kohsuke/stapler/export/Property.java

```diff
@@ -199,118 +198,127 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
 
         Class c = value.getClass();
 
-        Model model;
-        try {
-            model = owner.get(c, parent.type, name);
-        } catch (NotExportableException ex) {
-            if(STRING_TYPES.contains(c)) {
-                writer.value(value.toString());
-                return;
-            }
-            if(PRIMITIVE_TYPES.contains(c)) {
-                writer.valuePrimitive(value);
-                return;
-            }
-            Class act = c.getComponentType();
-            if (act !=null) { // array
-                Range r = pruner.getRange();
-                writer.startArray();
-                if (value instanceof Object[]) {
-                    // typical case
-                    for (Object item : r.apply((Object[]) value)) {
-                        writeBuffered(act, item, pruner, writer);
-                    }
-                } else {
-                    // more generic case
-                    int len = Math.min(r.max, Array.getLength(value));
-                    for (int i=r.min; i<len; i++) {
-                        writeBuffered(act, Array.get(value, i), pruner, writer);
-                    }
+        if (c.getAnnotation(ExportedBean.class) == null) {
+            handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
+        } else {
+            try {
+                Model model = owner.getOrNull(c, parent.type, name);
+                if (model == null) {
+                    throw new NotExportableException(c);
                 }
-                writer.endArray();
+                writer.startObject();
+                model.writeNestedObjectTo(value, pruner, writer);
+                writer.endObject();
+            } catch (NotExportableException ex) {
+                handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
                 return;
             }
-            if(value instanceof Iterable) {
-                writer.startArray();
-                Type expectedItemType = Types.getTypeArgument(expected, 0, null);
-                for (Object item : pruner.getRange().apply((Iterable) value)) {
-                    writeBuffered(expectedItemType, item, pruner, writer);
+        }
+
+        try {
+            writer.type(expected, value.getClass());
```

Comment:
This looks like 🐛 `writer.type()` should be called before writing the object on line 209.